python/cuml/metrics/kl_divergence.pyx

```diff
@@ -0,0 +1,101 @@
+#
+# Copyright (c) 2021, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import cupy as cp
+import numpy as np
+import cuml.internals
+from cuml.common.input_utils import determine_array_type
+from cuml.common import (input_to_cuml_array, CumlArray, logger)
+from libc.stdint cimport uintptr_t
+from cuml.raft.common.handle cimport handle_t
+from cuml.raft.common.handle import Handle
+
+cdef extern from "cuml/metrics/metrics.hpp" namespace "ML::Metrics":
+    double kl_divergence(const handle_t &handle,
+                         const double *y,
+                         const double *y_hat,
+                         int n) except +
+    float kl_divergence(const handle_t &handle,
+                        const float *y,
+                        const float *y_hat,
+                        int n) except +
+
+
+@cuml.internals.api_return_any()
+def cython_kl_divergence(P, Q, handle=None, convert_dtype=True):
+    """
+    Calculates the "Kullback-Leibler" Divergence
+    The KL divergence tells us how well the probability distribution Q
+    approximates the probability distribution P
+    It is often also used as a 'distance metric' between two probablity
+    ditributions (not symmetric)
+
+        Parameters
+        ----------
+        P : Array of probabilities corresponding to distribution P
+           Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+           ndarray, cuda array interface compliant array like CuPy.
+
+        Q : Array of probabilities corresponding to distribution Q
+           Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+           ndarray, cuda array interface compliant array like CuPy.
+
+        handle : cuml.Handle
+
+        convert_dtype : bool, optional (default = True)
+            When set to True, the method will, when necessary, convert
+            Q to be the same data type as P if they differ. This
+            will increase memory used for the method.
+        Returns
+        -------
+        float
+           The KL Divergence value
+    """
+    handle = Handle() if handle is None else handle
+    cdef handle_t *handle_ = <handle_t*> <size_t> handle.getHandle()
+
+    P_m, n_features_p, _, dtype_p = \
+        input_to_cuml_array(P, check_dtype=[np.float32, np.float64])
```

Comment:
Use `convert_dtype` here also

Reply:
I think it's better to not convert here because both float and double are accepted on the C++ side.

Reply:
That's alright, it won't convert if float or double are passed. It will convert if an int or long is passed. `convert_dtype = np.float32 if convert_dtype else None`